pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java

```diff
@@ -575,44 +587,59 @@ private void createDerivedColumnV1Indices(String column, FunctionEvaluator funct
         for (int j = 0; j < numArguments; j++) {
           inputValues[j] = valueReaders.get(j).getValue(i);
         }
-        Object outputValue = functionEvaluator.evaluate(inputValues);
-        outputValues[i] = outputValue;
-        if (outputValueType == null) {
+
+        Object outputValue = null;
+        try {
+          outputValue = functionEvaluator.evaluate(inputValues);
+        } catch (Exception e) {
+          if (!errorOnFailure) {
+            LOGGER.debug("Encountered an exception while evaluating function {} for derived column {} with "
+                    + "arguments: {}", functionEvaluator, column, Arrays.toString(inputValues), e);
+            functionEvaluateErrorCount++;
+            if (functionEvalError == null) {
+              functionEvalError = e;
+              inputValuesWithError = Arrays.copyOf(inputValues, inputValues.length);
+            }
+          } else {
+            throw e;
+          }
+        }
+
+        if (outputValue == null) {
+          outputValue = fieldSpec.getDefaultNullValue();
```

Comment:
Should we consider not setting `outputValue`, but just leave it as `null`? In the following handling, we can simply check `null` and fill with default value

Reply:
I think it's better to handle it here in a single place rather than adding `null` handling in multiple branches for each type? Any particular reason for doing it that way instead?

Reply:
Currently we rely on `instanceof` check to handle the case of default value. I was thinking to explicitly check for `null`, and always do the type conversion for non-null value (closer to the current logic), but the new code can avoid unnecessary type conversions, which is even better